authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,726 @@
+# Authentication via Social Networks
+
+Omniauth schema:
+![alt text](omniauth-schema.png "Omniauth schema")
+
+```ruby
+# Gemfile
+...
+gem "omniauth-facebook"
+gem "omniauth-google-oauth2"
+...
+```
+
+```ruby
+# app/controllers/identities_controller.rb
+class IdentitiesController < ApplicationController
+  before_action :authenticate_user!
+
+  expose(:identities) { current_user.identities }
+  expose(:identity)
+
+  def destroy
+    action = indentity.destroy ? :notice : :alert
+    flash[action] = t "flash.actions.destroy.#{action}", resource_name: Identity.model_name.human
+    redirect_to edit_user_registration_url
+  end
+end
+```
+
+```ruby
+# app/controllers/omniauth_callbacks_controller.rb
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  include OmniauthHelper
+
+  Identity::PROVIDERS.each do |provider|
+    define_method(provider) do
+      show_verification_notice and return unless auth_verified?
+
+      current_user ? connect_identity : process_sign_in
+    end
+  end
+
+  private
+
+  def show_verification_notice
+    redirect_to root_path, flash: { error: t("omniauth.verification.failure", kind: provider_name(auth.provider)) }
+  end
+
+  def auth_verified?
+    AuthVerificationPolicy.new(auth).verified?
+  end
+
+  def auth
+    request.env["omniauth.auth"]
+  end
+
+  def connect_identity
+    ConnectIdentity.new(current_user, auth).call
+    redirect_to edit_user_registration_path
+  end
+
+  def process_sign_in
+    user = FetchOauthUser.new(auth).call
+    sign_in_and_redirect user, event: :authentication
+  end
+end
+```
+
+```ruby
+# app/helpers/omniauth_helper.rb
+module OmniauthHelper
+  def provider_name(provider)
+    t "active_record.attributes.identity.provider_name.#{provider}"
+  end
+end
+```
+
+```ruby
+# app/interactors/connect_identity.rb
+class ConnectIdentity
+  attr_reader :user, :auth
+  private :user, :auth
+
+  def initialize(user, auth)
+    @user = user
+    @auth = auth
+  end
+
+  def call
+    update_or_create_identity
+    confirm_user
+  end
+
+  private
+
+  def update_or_create_identity
+    identity.present? ? update_identity : create_identity
+  end
+
+  def identity
+    @identity ||= Identity.from_omniauth(auth)
+  end
+
+  def update_identity
+    identity.update_attribute(:user, user)
+  end
+
+  def create_identity
+    user.identities.create!(provider: auth.provider, uid: auth.uid)
+  end
+
+  def confirm_user
+    user.confirm if user.email == auth.info.email
+  end
+end
+```
+
+```ruby
+# app/interactors/create_user_from_auth.rb
+class CreateUserFromAuth
+  attr_reader :auth
+  private :auth
+
+  def initialize(auth)
+    @auth = auth
+  end
+
+  def call
+    user = User.new(user_params)
```

Comment:
Also `auth` hashes for each provider are **different**. I suggest to have parsers for each provider to properly get data from `auth`